### PR TITLE
PC-139 Upgrade Morphia to v2.4.7; Build with Java 21 and target Java 11

### PR DESCRIPTION
**Pull request by rdlf0**
This PR includes:
- An upgrade of Morphia to the latest currently available version **2.4.7**.
- An upgrade of the MongoDB driver to version **4.11.1**
- An upgrade of the version of Java used to build the project to **21**. For the moment, the targeted version is **11** for most of the modules, except for the API and the client module, where we still target **8** for backward compatibility with older services. Thanks to this upgrade, some new language features were utilized, like the new immutable collections and methods, instead of Guava.
- Bumped the minor version of the project in order to emphasize some details that need to be kept in mind like:
  - `MongoStoreProvider.addCredential()` was renamed to `MongoStoreProvider.setCredential()`
  - `MongoStoreProvider.setDatabase()` is deprecated in favor of the [new Morphia configuration file property](https://morphia.dev/morphia/2.4/configuration.html)
  - `MongoStoreProvider` now has a `setClusterConnectionMode()` that defaults to `ClusterConnectionMode.SINGLE` and can be used to change the connection mode
  - The `MongoClientOptions.Builder.legacyDefaults()` method [no longer exists](https://github.com/mongodb/mongo-java-driver/blob/3.12.x/driver-legacy/src/main/com/mongodb/MongoClientOptions.java#L1715-L1726) in the current driver version, and should be replaced by manually setting `connectionsPerHost()` and `writeConcern()`
  - The `MongoClientOptions.Builder.socketFactory()` method [no longer exists](https://github.com/mongodb/mongo-java-driver/blob/3.12.x/driver-legacy/src/main/com/mongodb/MongoClientOptions.java#L1546-L1558) in the current driver version, and should be replaced by manually setting `sslEnabled()` and `sslContext()`
  - The new `DateTimeCodec` that replaces the `DateTimeConverter` can handle the timestamps used both in the new and the old database entries, while the old converter can't work with the dates in the new entries. So once upgraded, in case of a revert, the service won't be able to work with the documents created during the upgraded period.

Tested by building locally and verifying the build is successful and tests pass.
Ran the service locally and went through all available endpoints, verifying their behavior has remained unchanged.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -81,7 +80,7 @@
         <pluginManagement>
             <plugins>
                 <plugin>
-                    <groupId>com.coveo</groupId>
+                    <groupId>com.spotify.fmt</groupId>
                     <artifactId>fmt-maven-plugin</artifactId>
                     <version>${fmt-maven-plugin.version}</version>
                     <configuration>
```
Comment (MatthewRyanRead) [inline pom.xml:86]: Can we remove `skipSortingImports` now?

Comment (rdlf0) [reply pom.xml:86]: I was about to do it, but then realized we will need to undo the imports in the autogenerated files every time, unless we find a way to exclude them somehow.

File: alchemy-api/pom.xml [hunk 2]
```diff
@@ -66,4 +66,17 @@
       <scope>test</scope>
     </dependency>
   </dependencies>
+
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-compiler-plugin</artifactId>
+        <configuration>
```
[the remark this answers is not in the document]
Comment (rdlf0) [reply alchemy-api/pom.xml:75]: I have to see where I copied it from, so that I change it there, too 🙃

File: pom.xml (hunk 1 above)
Comment (MatthewRyanRead) [reply pom.xml:86]: The autogenerated files are ignored because of the config in the module though, yeah?

Comment (rdlf0) [reply pom.xml:86]: I was hoping to find a way to utilize the `filesNamePattern` config option so that we apply it on everything except them.

**Review by MatthewRyanRead — APPROVED** [review]

File: pom.xml (hunk 1 above)
Comment (MatthewRyanRead) [reply pom.xml:86]: Gotcha, yeah we can look into that for a future improvement

File: alchemy-core/src/main/java/io/rtr/alchemy/models/Allocations.java [hunk 3]
```diff
@@ -120,16 +113,16 @@ private void mergeAdjacentAllocations() {
...
      */
-    public void allocate(Treatment treatment, int size) {
-        Preconditions.checkArgument(
-                getUnallocatedSize() >= size,
-                "not enough free bins to allocate treatment %s with size %s given %s unallocated bin(s)",
-                treatment.getName(),
-                size,
-                getUnallocatedSize());
+    public void allocate(final Treatment treatment, final int size) {
+        if (getUnallocatedSize() < size) {
```
Comment (heeseungp) [inline alchemy-core/src/main/java/io/rtr/alchemy/models/Allocations.java:117]: Is this the right logic? I see unallocatedSize() >= size in the previous version

Comment (rdlf0) [reply alchemy-core/src/main/java/io/rtr/alchemy/models/Allocations.java:117]: Yes, `Preconditions` need "how it should be" as an argument.

**Review by heeseungp — APPROVED** [review]

**Pull request merged**